### PR TITLE
Store Exception for on headers and onbody callback

### PR DESCRIPTION
**Pull request by waahm7**
*Description of changes:*
- Store callback exceptions in S3 on_headers and on_body callback so that we rethrow it in the on_done callback.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: awscrt/s3.py [hunk 1]
```diff
@@ -529,11 +531,19 @@ def __init__(
...
+            except Exception as e:
+                self._python_callback_exception = e
+                raise
 
     def _on_body(self, chunk, offset):
         if self._on_body_cb:
-            self._on_body_cb(chunk=chunk, offset=offset)
+            try:
+                self._on_body_cb(chunk=chunk, offset=offset)
+            except Exception as e:
```
Comment (graebm) [inline awscrt/s3.py:544]: BaseException

File: awscrt/s3.py [hunk 2]
```diff
@@ -545,11 +555,14 @@ def _on_finish(self, error_code, status_code, error_headers, error_body):
 
         error = None
         if error_code:
-            error = awscrt.exceptions.from_code(error_code)
 
+            error = awscrt.exceptions.from_code(error_code)
+            if (error.name == "AWS_ERROR_CRT_CALLBACK_EXCEPTION"
```
Comment (graebm) [inline awscrt/s3.py:560]: first check that it's an AwsCrtError. it's possible for this to return a built-in python exception, which doesn't have a `.name` property

File: source/s3_meta_request.c [hunk 3]
```diff
@@ -103,6 +103,11 @@ static int s_s3_request_on_headers(
...
         PyErr_WriteUnraisable(request_binding->py_core);
         goto done;
+    } else if (result == Py_False) {
+        /* _on_headers stored the exception to throw later, so we don't have to do anything */
+        PyErr_Clear();
+        Py_DECREF(result);
+        goto done;
     }
     Py_DECREF(result);
     error = false;
```
Comment (graebm) [inline source/s3_meta_request.c:113]:
```suggestion
    }
    /* If user's callback raises an exception, _S3RequestCore._on_headers
     * stores it to throw later and returns False */
    error = (result == Py_False);
    Py_DECREF(result);
```

File: source/s3_meta_request.c [hunk 4]
```diff
@@ -176,6 +181,11 @@ static int s_s3_request_on_body(
     if (!result) {
         PyErr_WriteUnraisable(request_binding->py_core);
         goto done;
+    } else if (result == Py_False) {
+        /* _on_body stored the exception to throw later, so we don't have to do anything */
+        PyErr_Clear();
+        Py_DECREF(result);
+        goto done;
```
Comment (graebm) [inline source/s3_meta_request.c:188]: same

File: test/test_s3.py [hunk 5]
```diff
@@ -617,6 +617,48 @@ def test_multipart_upload_with_invalid_request(self):
...
         put_body_stream.close()
 
+    def test_on_headers_callback_failure(self):
+        request = self._get_object_request(self.get_test_object_path)
+
+        s3_client = s3_client_new(False, self.region, 5 * MB)
+        s3_request = s3_client.make_request(
+            request=request,
+            type=S3RequestType.GET_OBJECT,
+            on_headers=lambda status_code, headers: (_ for _ in ()).throw(RuntimeError("Error in on_headers")),
```
Comment (graebm) [inline test/test_s3.py:627]: trivial: `(_ for _ in ()).throw` <-- wut

might I suggest a local function:
```
def _explode(**kwargs):
    raise RuntimeError("Error in on_headers")

...
    on_headers=_explode,
```

**Review by graebm — APPROVED** [review]
fix and ship

File: test/test_s3.py (hunk 5 above)
Comment (waahm7) [reply test/test_s3.py]: Thanks, updated.

**Pull request merged**